rustls/src/crypto.rs

```diff
@@ -18,8 +21,58 @@ pub trait CryptoProvider: Send + Sync + 'static {
     fn verify_equal_ct(a: &[u8], b: &[u8]) -> bool;
 }
 
-pub(crate) enum KeyExchangeError {
+/// An in-progress key exchange over a [SupportedGroup].
+pub trait KeyExchange: Sized + Send + Sync + 'static {
+    /// The supported group the key exchange is operating over.
+    type SupportedGroup: SupportedGroup;
+
+    /// Start a key exchange using the [NamedGroup] if it is a suitable choice
+    /// based on the groups supported.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the [NamedGroup] is not supported, or if a key exchange
+    /// can't be started (see [KeyExchange#start]).
+    fn choose(
+        name: NamedGroup,
+        supported: &[&'static Self::SupportedGroup],
+    ) -> Result<Self, KeyExchangeError>
+    where
+        Self: Sized;
+
+    /// Start a key exchange using the [SupportedGroup]. This will prepare an ephemeral
+    /// secret key in the supported group, and a corresponding public key. The key exchange
+    /// must be completed by calling [KeyExchange#complete].
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if key generation fails.
+    fn start(skxg: &'static Self::SupportedGroup) -> Result<Self, GetRandomFailed>
+    where
+        Self: Sized;
+
+    /// Completes the key exchange, given the peer's public key.
+    ///
+    /// The shared secret is passed into the closure passed down in `f`, and the result of calling
+    /// `f` is returned to the caller.
+    fn complete<T>(self, peer: &[u8], f: impl FnOnce(&[u8]) -> Result<T, ()>) -> Result<T, Error>;
+
+    /// Return all supported key exchange groups.
+    fn all_kx_groups() -> &'static [&'static Self::SupportedGroup];
```

Comment:
In this case I'd move `all_kx_groups()` below `pub_key()`, since it doesn't take `&self` (sort of like a constant).